sktime/forecasting/neuralforecast.py

```diff
@@ -0,0 +1,343 @@
+        instantiated train loss class from losses collection [5]_
+    valid_loss : pytorch module (default=None)
+        instantiated validation loss class from losses collection [5]_
+    max_steps : int (default=1000)
+        maximum number of training steps
+    learning_rate : float (default=1e-3)
+        learning rate between (0, 1)
+    num_lr_decays : int (default=-1)
+        number of learning rate decays, evenly distributed across max_steps
+    early_stop_patience_steps : int (default=-1)
+        number of validation iterations before early stopping
+    val_check_steps : int (default=100)
+        number of training steps between every validation loss check
+    batch_size : int (default=32)
+        number of different series in each batch
+    valid_batch_size : typing.Optional[int] (default=None)
+        number of different series in each validation and test batch
+    scaler_type : str (default="robust")
+        type of scaler for temporal inputs normalization
+    random_seed : int (default=1)
+        random_seed for pytorch initializer and numpy generators
+    num_workers_loader : int (default=0)
+        workers to be used by `TimeSeriesDataLoader`
+    drop_last_loader : bool (default=False)
+        whether `TimeSeriesDataLoader` drops last non-full batch
+    trainer_kwargs : dict (default=None)
+        keyword trainer arguments inherited from PyTorch Lighning's trainer [6]_
+
+    Notes
+    -----
+    * If ``loss`` is unspecified, MAE is used as the loss function for training.
+    * Only ``futr_exog_list`` will be considered as exogenous variables.
+
+    Examples
+    --------
+    >>>
+    >>> # importing necessary libraries
+    >>> from sktime.datasets import load_longley
+    >>> from sktime.forecasting.neuralforecast import NeuralForecastRNN
+    >>> from sktime.split import temporal_train_test_split
+    >>>
+    >>> # loading the Longley dataset and splitting it into train and test subsets
+    >>> y, X = load_longley()
+    >>> y_train, y_test, X_train, X_test = temporal_train_test_split(y, X, test_size=4)
+    >>>
+    >>> # creating model instance configuring the hyperparameters
+    >>> model = NeuralForecastRNN(  # doctest: +SKIP
+    ...     "A-DEC", futr_exog_list=["ARMED", "POP"], max_steps=5
+    ... )
+    >>>
+    >>> # fitting the model
+    >>> model.fit(y_train, X=X_train, fh=[1, 2, 3, 4])  # doctest: +SKIP
+    Seed set to 1
+    Epoch 4: 100%|█| 1/1 [00:00<00:00, 42.85it/s, v_num=870, train_loss_step=0.589, train_loss_epoc
+    NeuralForecastRNN(freq='A-DEC', futr_exog_list=['ARMED', 'POP'], max_steps=5)
+    >>>
+    >>> # getting point predictions
+    >>> model.predict(X=X_test)  # doctest: +SKIP
+    Predicting DataLoader 0: 100%|██████████████████████████████████| 1/1 [00:00<00:00, 198.64it/s]
+    1959    66241.984375
+    1960    66700.125000
+    1961    66550.195312
+    1962    67310.007812
+    Freq: A-DEC, Name: TOTEMP, dtype: float64
+    >>>
```

Comment:
This is how I visualised end users will use it. Is that what you are asking?

> question: how would we test this in the intended end state?

What tests do you suggest I should add?

Reply:
I mean not concrete tests but dependencies. That is, how does the end state of `pyproject.toml` look like?

Reply:
I completely misunderstood then.

I am open for suggestions. Here's what I already have though of.

1. I can leave as it is in `forecasting` extra, as I'd guess a lot of people may be interested in using `neuralforecast` models, anything automl (they have AutoRNN, AutoTFT, etc. quite a few) seems to be popular.
2. I can move to `dl`, but that means it will not be covered by tests once we retire old CI. I can add a new job in CI specifically to test `dl` extra, because I guess it needs to be tested not just in `forecasting` but `network` and `regression` as well. Please correct me if that's a wrong idea.

I am of course biased, but I'm tempted to do (1). An excuse is mainly that it'll allow us to have `ray[tune]` as well, which may invite others to contribute a new tuner similar to what @hazrulakmal did for skopt.

In any case, probably I should remove it from `all_extras` and `all_extras_pandas2` at least?